### PR TITLE
Nodes can generate schema only reports

### PR DESCRIPTION
**Pull request by jaroslaw-osmanski**

### CODE REVIEW

File: project/Dependencies.scala [hunk 1]
```diff
@@ -33,6 +33,7 @@ object Version {
 
   val spark = sys.props.getOrElse("SPARK_VERSION", "2.1.1")
   val (scala, hadoop, akka, sprayRoutingLib) = spark match {
+    case "2.2.0"  => ("2.11.8", "2.7", "2.4.9", "routing-shapeless2")
     case "2.1.0" | "2.1.1" => ("2.11.8", "2.7", "2.4.9", "routing-shapeless2")
```
Comment (witold-jedrzejewski-codilime) [inline project/Dependencies.scala]: why not add it to next line? case "2.1.0" | "2.1.1" | "2.2.0"

File: seahorse-workflow-executor/build.sbt [hunk 2]
```diff
@@ -22,23 +22,51 @@ lazy val settingsForNotPublished = CommonSettingsPlugin.assemblySettings ++
...
 lazy val sparkUtils = sparkVersion match {
   case "2.0.0" | "2.0.1" | "2.0.2" =>
     val sparkUtils2_0_x = project in file("sparkutils2.0.x") settings settingsForPublished
     sparkUtils2_0_x
   case "2.1.0" | "2.1.1" =>
     val sparkUtils2_1_0 = project in file("sparkutils2.1.x") settings settingsForPublished
     sparkUtils2_1_0
+  case "2.2.0" =>
+    val sparkUtils2_1_0 = project in file("sparkutils2.2.x") settings settingsForPublished
+    sparkUtils2_1_0
```
Comment (witold-jedrzejewski-codilime) [inline seahorse-workflow-executor/build.sbt]: rename to 2_2_0

File: remote_notebook/code/sparkr_kernel/kernel_init.R [hunk 3]
```diff
@@ -16,7 +16,7 @@ assign("sc", get(".sparkRjsc", envir = SparkR:::.sparkREnv), envir = .GlobalEnv)
 sparkSQLSession <- SparkR:::callJMethod(entryPoint, "getNewSparkSQLSession")
 
 sparkVersion <- SparkR:::callJMethod(sc, "version")
-if (sparkVersion %in% c("2.0.0", "2.0.1", "2.0.2", "2.1.0", "2.1.1")) {
+if (sparkVersion %in% c("2.0.0", "2.0.1", "2.0.2", "2.1.0", "2.1.1", "2.2.0", "2.2.1")) {
```
Comment (witold-jedrzejewski-codilime) [inline remote_notebook/code/sparkr_kernel/kernel_init.R]: do we add support for 2.2.0 or for 2.2.1 as well? it differs from file to file

File: seahorse-workflow-executor/deeplang/src/main/scala/ai/deepsense/deeplang/DOperation.scala [hunk 4]
```diff
@@ -98,9 +102,39 @@ abstract class DOperation extends Operation
   def inferGraphKnowledgeForInnerWorkflow(context: InferContext): GraphKnowledge = GraphKnowledge()
 
   def typeTag[T : ru.TypeTag]: ru.TypeTag[T] = ru.typeTag[T]
+
+  val reportType: ChoiceParam[ReportType] = ChoiceParam[ReportType](
+    name = "report type",
+    description = Some(""))
```
Comment (witold-jedrzejewski-codilime) [inline seahorse-workflow-executor/deeplang/src/main/scala/ai/deepsense/deeplang/DOperation.scala]: why empty description? This is something that gets rendered in the tooltip. Should be something informative, not empty

File: seahorse-workflow-executor/deeplang/src/main/scala/ai/deepsense/deeplang/DOperation.scala [hunk 5]
```diff
@@ -98,9 +102,39 @@ abstract class DOperation extends Operation
...
+
+    object ReportParam {
+      case class SchemaOnly() extends ReportType {
+        override val name: String = "Report schema only"
+        override val params: Array[Param[_]] = Array()
+      }
+
+      case class Extended() extends ReportType {
+        override val name: String = "Report schema, count and for small reports content"
+        override val params: Array[Param[_]] = Array()
```
Comment (witold-jedrzejewski-codilime) [inline seahorse-workflow-executor/deeplang/src/main/scala/ai/deepsense/deeplang/DOperation.scala]: Those descriptions are accurate only for dataframes. However, operations are producing reports for many different doperables

File: seahorse-workflow-executor/scalastyle-config.xml [hunk 6]
```diff
@@ -22,7 +22,7 @@
     <parameters>
       <parameter name="regex">true</parameter>
       <parameter name="header"><![CDATA[/\*\*
- \* Copyright 201[567] deepsense.ai \(CodiLime, Inc\)
+ \* Copyright 201[56789] deepsense.ai \(CodiLime, Inc\)
```
Comment (witold-jedrzejewski-codilime) [inline seahorse-workflow-executor/scalastyle-config.xml]: see you in 2 years

File: build/build_and_run_tests.sh [hunk 7]
```diff
@@ -32,6 +32,7 @@ run_tests() {
   sbt clean
   sbt -DSPARK_VERSION=$1 test ds-it
 }
+run_tests 2.2.0
```
Comment (witold-jedrzejewski-codilime) [inline build/build_and_run_tests.sh]: This PR mixes the feature described in its title with upgrade to new spark. Can it be separated?

File: seahorse-workflow-executor/deeplang/src/main/scala/ai/deepsense/deeplang/params/ParamMap.scala [hunk 8]
```diff
@@ -52,6 +52,10 @@ case class ParamMap private[deeplang] (private val map: mutable.Map[Param[Any],
     map.remove(param.asInstanceOf[Param[Any]]).asInstanceOf[Option[T]]
   }
 
+  def -[T](param: Param[T]): ParamMap = {
```
Comment (witold-jedrzejewski-codilime) [inline seahorse-workflow-executor/deeplang/src/main/scala/ai/deepsense/deeplang/params/ParamMap.scala]: where this is used?

File: seahorse-workflow-executor/deeplang/src/main/scala/ai/deepsense/deeplang/doperables/dataframe/DataFrame.scala [hunk 9]
```diff
@@ -72,8 +72,11 @@ case class DataFrame private[dataframe] (
     DataFrame.fromSparkDataFrame(newSparkDataFrame)
   }
 
-  override def report: Report = {
-    DataFrameReportGenerator.report(sparkDataFrame)
+  override def report(extended: Boolean): Report = {
+    extended match {
+      case true => DataFrameReportGenerator.report (sparkDataFrame)
```
Comment (witold-jedrzejewski-codilime) [inline seahorse-workflow-executor/deeplang/src/main/scala/ai/deepsense/deeplang/doperables/dataframe/DataFrame.scala]: remove space

File: seahorse-workflow-executor/scalastyle-config.xml (hunk 6 above)
Comment (jaroslaw-osmanski) [reply seahorse-workflow-executor/scalastyle-config.xml]: Actually 8 months

File: seahorse-workflow-executor/deeplang/src/main/scala/ai/deepsense/deeplang/params/ParamMap.scala (hunk 8 above)
Comment (jaroslaw-osmanski) [reply seahorse-workflow-executor/deeplang/src/main/scala/ai/deepsense/deeplang/params/ParamMap.scala]: Removed

File: seahorse-workflow-executor/deeplang/src/main/scala/ai/deepsense/deeplang/doperables/dataframe/DataFrame.scala (hunk 9 above)
Comment (jaroslaw-osmanski) [reply seahorse-workflow-executor/deeplang/src/main/scala/ai/deepsense/deeplang/doperables/dataframe/DataFrame.scala]: Removed

File: seahorse-workflow-executor/deeplang/src/main/scala/ai/deepsense/deeplang/DOperation.scala [hunk 10]
```diff
@@ -98,9 +102,37 @@ abstract class DOperation extends Operation
   def inferGraphKnowledgeForInnerWorkflow(context: InferContext): GraphKnowledge = GraphKnowledge()
 
   def typeTag[T : ru.TypeTag]: ru.TypeTag[T] = ru.typeTag[T]
+
+  val reportType: ChoiceParam[ReportType] = ChoiceParam[ReportType](
+    name = "report type",
+    description = Some("Reports extensiveness"))
```
Comment (witold-jedrzejewski-codilime) [inline seahorse-workflow-executor/deeplang/src/main/scala/ai/deepsense/deeplang/DOperation.scala]: Maybe something like:
"What kind of report should be generated"

File: seahorse-workflow-executor/deeplang/src/main/scala/ai/deepsense/deeplang/DOperation.scala [hunk 11]
```diff
@@ -98,9 +102,37 @@ abstract class DOperation extends Operation
...
+
+  def setReportType(value: ReportType): this.type = set(reportType, value)
 }
 
 object DOperation {
   type Id = models.Id
   val Id = models.Id
+
+  object ReportParam {
+    case class SchemaOnly() extends ReportType {
```
Comment (witold-jedrzejewski-codilime) [inline seahorse-workflow-executor/deeplang/src/main/scala/ai/deepsense/deeplang/DOperation.scala]: rename to Minimal? or Simplified?

Comment (jaroslaw-osmanski) [reply seahorse-workflow-executor/deeplang/src/main/scala/ai/deepsense/deeplang/DOperation.scala]: Minimal reports are generated when there are over 100 rows.

File: seahorse-workflow-executor/deeplang/src/main/scala/ai/deepsense/deeplang/DOperation.scala (hunk 10 above)
Comment (jaroslaw-osmanski) [reply seahorse-workflow-executor/deeplang/src/main/scala/ai/deepsense/deeplang/DOperation.scala]: "What kind of report should be generated" looks like just synonym for report type.

File: seahorse-workflow-executor/deeplang/src/main/scala/ai/deepsense/deeplang/DOperation.scala (hunk 11 above)
Comment (witold-jedrzejewski-codilime) [reply seahorse-workflow-executor/deeplang/src/main/scala/ai/deepsense/deeplang/DOperation.scala]: Well, the name of this param says "Minimal report". Let's figure out some sensible name and have name of this class match the 'name' field.

File: seahorse-workflow-executor/deeplang/src/main/scala/ai/deepsense/deeplang/DOperation.scala (hunk 10 above)
Comment (witold-jedrzejewski-codilime) [reply seahorse-workflow-executor/deeplang/src/main/scala/ai/deepsense/deeplang/DOperation.scala]: I feel that it adds new information. But I'm open for other ideas. You can have a look at other parameters descriptions in seahorse to get a feel of what we are aiming at.

Another idea:
have this param as boolean (yes/no) and have
* name: generate rich reports
* description: If this is enabled, extensive reports for output entities are generated. Note that this involve additional computation.

Comment (jaroslaw-osmanski) [reply seahorse-workflow-executor/deeplang/src/main/scala/ai/deepsense/deeplang/DOperation.scala]: We might introduce new report types and then old workflows would be incompatible.

**Comment by jaroslaw-osmanski** [comment]
I don't understand what actual usage of the param means?
Test for the param:
1. testing DataFrame.report(extend=false)
2. TransformerAsOperation checks if param is being set
3. End to End test with workflow
What more can be done?

**Pull request merged**